File: llmem/paths.py
"""Path resolution for llmem data directories.

All hardcoded ~/.lobsterdog/ paths are replaced with ~/.config/llmem/
(with LMEM_HOME env var override). Backward compatibility:
~/.lobsterdog/ is checked first for migration.
"""

import logging
import os
import shutil
from pathlib import Path

log = logging.getLogger(__name__)

# Maximum allowed components in LMEM_HOME path (prevents /../../../etc tricks)
_MAX_PATH_DEPTH = 10

# System directories that should never be used as llmem data locations.
# Shared between _validate_home_path, _validate_write_path, and
# OpenCodeAdapter.__init__ to prevent DRY violations.
BLOCKED_SYSTEM_PREFIXES = (
    "/etc",
    "/var",
    "/sys",
    "/proc",
    "/dev",
    "/boot",
    "/root",
    "/sbin",
    "/bin",
    "/usr/sbin",
    "/usr/bin",
)


def _validate_home_path(path: Path, source: str) -> Path:
    """Validate that a home path is safe to use.

    Checks (in order):
    - Must not contain '..' traversal components (checked before resolve)
    - Must not target sensitive system directories (checked before symlink
      check since is_symlink() requires stat access which may fail)
    - Must not be a symlink itself (prevents symlink escalation)
    - Must not exceed a reasonable path depth

    Args:
        path: The candidate home path.
        source: Description of where this path came from (for error messages).

    Returns:
        The resolved, validated path.

    Raises:
        ValueError: If the path is unsafe.
    """
    # Check traversal BEFORE resolving (resolve eliminates ..)
    if ".." in str(path):
        raise ValueError(f"llmem: paths: {source} contains '..' traversal: {path}")

    resolved = path.resolve()

    # Block obvious system directories — checked before symlink check
    # because is_symlink() requires stat access which may fail for
    # inaccessible paths like /root
    for prefix in BLOCKED_SYSTEM_PREFIXES:
        if str(resolved).startswith(prefix):
            raise ValueError(
                f"llmem: paths: {source} targets a system directory: {resolved}"
            )

    # Must not be a symlink itself (prevents symlink escalation)
    # If we can't stat the path (permission denied), treat it as unsafe
    try:
        if path.is_symlink():
            raise ValueError(
                f"llmem: paths: {source} is a symlink (not allowed): {path}"
            )
    except OSError:
        raise ValueError(
            f"llmem: paths: {source} cannot be accessed (permission denied): {path}"
        )

    # Must not exceed a reasonable path depth
    parts = resolved.parts
    if len(parts) > _MAX_PATH_DEPTH:
        raise ValueError(
            f"llmem: paths: {source} path too deep ({len(parts)} components): {resolved}"
        )

    return resolved


def get_llmem_home() -> Path:
    """Return the base directory for all llmem data.

    Resolution order:
    1. LMEM_HOME env var (if set and non-empty after stripping)
    2. ~/.config/llmem/ (if it exists)
    3. ~/.lobsterdog/ (backward compat — if it exists and ~/.config/llmem/ doesn't)
    4. ~/.config/llmem/ (default, even if it doesn't exist yet)

    Returns:
        Path — never None. The path may not exist on disk yet.

    Raises:
        ValueError: If LMEM_HOME is set but points to an unsafe location.
    """
    env_val = os.environ.get("LMEM_HOME", "").strip()
    if env_val:
        candidate = Path(env_val)
        return _validate_home_path(candidate, "LMEM_HOME")

    new_path = Path.home() / ".config" / "llmem"
    old_path = Path.home() / ".lobsterdog"

    if new_path.exists():
        return new_path
    if old_path.exists():
        log.info(
            "paths: using legacy ~/.lobsterdog/ directory; "
            "run migrate_from_lobsterdog() to move to ~/.config/llmem/"
        )
        return old_path

    return new_path


def _validate_write_path(path: Path, label: str) -> Path:
    """Validate that a write target path is safe.

    Checks:
    - Must not contain '..' traversal components (checked before resolve)
    - Must not target protected system directories
    - Must not be a symlink itself

    Does NOT require the path to be within the llmem home directory —
    users may configure custom output paths. This function prevents
    clearly dangerous writes, not all writes outside the default home.

    Args:
        path: The candidate write path.
        label: Description of the file being written (for error messages).

    Returns:
        The resolved, validated path.

    Raises:
        ValueError: If the path is unsafe.
    """
    # Check traversal BEFORE resolving (resolve eliminates ..)
    if ".." in str(path):
        raise ValueError(f"llmem: paths: {label} path contains '..' traversal: {path}")

    resolved = path.resolve()

    # Block system directories
    for prefix in BLOCKED_SYSTEM_PREFIXES:
        if str(resolved).startswith(prefix):
            raise ValueError(
                f"llmem: paths: {label} path targets a protected directory: {resolved}"
            )

    # Must not be a symlink itself
    # If we can't stat the path (permission denied), treat it as unsafe
    try:
        if path.is_symlink():
            raise ValueError(
                f"llmem: paths: {label} path is a symlink (not allowed for write targets): {path}"
            )
    except OSError:
        raise ValueError(
            f"llmem: paths: {label} path cannot be accessed (permission denied): {path}"
        )

    return resolved


def validate_session_id(session_id: str) -> str:
    """Validate that a session ID is safe to use in filesystem paths.

    Rejects session IDs that contain path separators or traversal sequences,
    which could allow writing files outside the intended context directory.

    Args:
        session_id: The session ID to validate.

    Returns:
        The validated session ID string (unchanged if valid).

    Raises:
        ValueError: If the session ID contains '/', '\\', or '..' sequences.
    """
    if not session_id:
        raise ValueError("llmem: paths: session_id must not be empty")
    if "/" in session_id:
        raise ValueError(
            f"llmem: paths: session_id contains '/' (path traversal risk): {session_id!r}"
        )
    if "\\" in session_id:
        raise ValueError(
            f"llmem: paths: session_id contains '\\' (path traversal risk): {session_id!r}"
        )
    if ".." in session_id:
        raise ValueError(
            f"llmem: paths: session_id contains '..' (path traversal risk): {session_id!r}"
        )
    return session_id


def get_config_path() -> Path:
    """Return the path to config.yaml.

    Returns:
        get_llmem_home() / "config.yaml". No preconditions.
    """
    return get_llmem_home() / "config.yaml"


def get_db_path() -> Path:
    """Return the path to the memory database.

    Returns:
        get_llmem_home() / "memory.db". No preconditions.
    """
    return get_llmem_home() / "memory.db"


def get_dream_diary_path() -> Path:
    """Return the path to the dream diary.

    Returns:
        get_llmem_home() / "dream-diary.md". No preconditions.
    """
    return get_llmem_home() / "dream-diary.md"


def get_proposed_changes_path() -> Path:
    """Return the path to the proposed changes file.

    Returns:
        get_llmem_home() / "proposed-changes.md". No preconditions.
    """
    return get_llmem_home() / "proposed-changes.md"


def get_context_dir() -> Path:
    """Return the path to the context directory.

    Returns:
        get_llmem_home() / "context". No preconditions.
    """
    return get_llmem_home() / "context"


def migrate_from_lobsterdog() -> bool:
    """Migrate data from ~/.lobsterdog/ to ~/.config/llmem/.

    Copies config.yaml, memory.db, dream-diary.md, proposed-changes.md,
    and context/ directory from ~/.lobsterdog/ to ~/.config/llmem/
    only if the source exists and the destination doesn't.
    Never deletes the source directory.

    Uses shutil.copy2 with follow_symlinks=False to avoid following
    symlinks — source symlinks are skipped, not followed.

    Returns:
        True if any migration was performed, False if no migration needed.
        Idempotent — calling twice does nothing on the second call.
    """
    old_home = Path.home() / ".lobsterdog"
    new_home = Path.home() / ".config" / "llmem"

    # Resolve to avoid symlink attacks on source/destination
    old_home = old_home.resolve()
    new_home = new_home.resolve()

    if not old_home.exists():
        return False
    if new_home.exists():
        return False

    migrated = False
    new_home.mkdir(parents=True, exist_ok=True)

    files_to_copy = [
        "config.yaml",
        "memory.db",
        "dream-diary.md",
        "proposed-changes.md",
    ]

    for filename in files_to_copy:
        src = old_home / filename
        dst = new_home / filename
        # Skip symlinks — only copy regular files
        if src.is_symlink():
            log.warning("paths: skipping symlink during migration: %s", src)
            continue
        if src.exists() and not dst.exists():
            try:
                shutil.copy2(str(src), str(dst), follow_symlinks=False)
                log.info(
                    "paths: migrated %s from ~/.lobsterdog/ to ~/.config/llmem/",
                    filename,
                )
                migrated = True
            except OSError as e:
                log.warning("paths: failed to migrate %s: %s", filename, e)

    # Copy context/ directory — skip symlinks
    src_context = old_home / "context"
    dst_context = new_home / "context"
    if src_context.is_symlink():
        log.warning(
            "paths: skipping symlink directory during migration: %s", src_context
        )
    elif src_context.is_dir() and not dst_context.exists():
        try:
            shutil.copytree(
                str(src_context),
                str(dst_context),
                symlinks=False,  # Don't preserve symlinks — copy referenced content
            )
            log.info(
                "paths: migrated context/ directory from ~/.lobsterdog/ to ~/.config/llmem/"
            )
            migrated = True
        except OSError as e:
            log.warning("paths: failed to migrate context/ directory: %s", e)

    return migrated

File: internal/paths/paths.go
// Package paths provides path resolution, validation, and migration for LLMem.
// All hardcoded paths use ~/.config/llmem/ (with LMEM_HOME override).
// Backward compatibility: ~/.lobsterdog/ is checked first for migration.
package paths

import (
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"strings"
)

// maxPathDepth limits path component count to prevent traversal attacks.
const maxPathDepth = 10

// blockedSystemPrefixes lists system directories that should never be used
// as LLMem data locations. Shared between ValidateHomePath and
// ValidateWritePath to prevent DRY violations.
var blockedSystemPrefixes = []string{
	"/etc",
	"/var",
	"/sys",
	"/proc",
	"/dev",
	"/boot",
	"/root",
	"/sbin",
	"/bin",
	"/usr/sbin",
	"/usr/bin",
}

// fmtErr wraps an error with the "llmem: paths:" domain prefix.
func fmtErr(format string, args ...any) error {
	return fmt.Errorf("llmem: paths: "+format, args...)
}

// GetHomeDir returns the base directory for all LLMem data.
// Resolution order:
// 1. LMEM_HOME env var (if set and non-empty after stripping)
// 2. ~/.config/llmem/ (if it exists)
// 3. ~/.lobsterdog/ (backward compat — if it exists and ~/.config/llmem/ doesn't)
// 4. ~/.config/llmem/ (default, even if it doesn't exist yet)
//
// Creates the directory with 0700 permissions if it doesn't exist.
func GetHomeDir() string {
	envVal := ""
	if v, ok := os.LookupEnv("LMEM_HOME"); ok {
		envVal = strings.TrimSpace(v)
	}
	if envVal != "" {
		resolved, err := ValidateHomePath(filepath.Clean(envVal), "LMEM_HOME")
		if err != nil {
			slog.Error("llmem: paths: invalid LMEM_HOME", "error", err)
			// Fall through to default
		} else {
			os.MkdirAll(resolved, 0700)
			return resolved
		}
	}

	homeDir, err := os.UserHomeDir()
	if err != nil {
		// Last resort
		return "/tmp/llmem"
	}

	newPath := filepath.Join(homeDir, ".config", "llmem")
	oldPath := filepath.Join(homeDir, ".lobsterdog")

	// If new path exists, use it
	if _, err := os.Stat(newPath); err == nil {
		return newPath
	}
	// If old path exists, use it (backward compat)
	if _, err := os.Stat(oldPath); err == nil {
		slog.Info("llmem: paths: using legacy ~/.lobsterdog/ directory; run MigrateFromLobsterdog() to move to ~/.config/llmem/")
		return oldPath
	}

	return newPath
}

// GetDBPath returns the path to the memory database (~/.config/llmem/memory.db).
func GetDBPath() string {
	return filepath.Join(GetHomeDir(), "memory.db")
}

// GetConfigPath returns the path to config.yaml (~/.config/llmem/config.yaml).
func GetConfigPath() string {
	return filepath.Join(GetHomeDir(), "config.yaml")
}

// GetDreamDiaryPath returns the path to the dream diary (~/.config/llmem/dream_diary.md).
func GetDreamDiaryPath() string {
	return filepath.Join(GetHomeDir(), "dream_diary.md")
}

// GetDreamReportPath returns the path to the dream report (~/.config/llmem/dream_report.html).
func GetDreamReportPath() string {
	return filepath.Join(GetHomeDir(), "dream_report.html")
}

// GetProposedChangesPath returns the path to the proposed-changes file (~/.config/llmem/proposed-changes.md).
func GetProposedChangesPath() string {
	return filepath.Join(GetHomeDir(), "proposed-changes.md")
}

// GetSkillDir returns the path to the skills directory (~/.config/llmem/skills/).
// Checks LMEM_HOME env var override, same as GetHomeDir().
func GetSkillDir() string {
	return filepath.Join(GetHomeDir(), "skills")
}

// GetContextDir returns the path to the context directory (~/.config/llmem/context/).
func GetContextDir() string {
	return filepath.Join(GetHomeDir(), "context")
}

// GetDirFromPath returns the parent directory of a file path.
// Returns empty string if the path has no directory component.
func GetDirFromPath(path string) string {
	return filepath.Dir(path)
}

// IsBlockedPath returns true if the resolved path targets a protected system directory.
// Uses prefix + '/' matching to avoid false positives — e.g. '/bin'
// should not match '/binary_search/data.db', but '/bin/' should.
func IsBlockedPath(resolved string) bool {
	for _, prefix := range blockedSystemPrefixes {
		if resolved == prefix || strings.HasPrefix(resolved, prefix+"/") {
			return true
		}
	}
	return false
}

// ValidateHomePath validates that a home path is safe to use.
// Checks (in order):
// - Must not contain '..' traversal components
// - Must not target sensitive system directories
// - Must not be a symlink itself
// - Must not exceed reasonable path depth
//
// Returns the resolved, validated path.
func ValidateHomePath(path, source string) (string, error) {
	// Check traversal BEFORE resolving (resolve eliminates ..)
	if strings.Contains(filepath.Clean(path), "..") {
		return "", fmtErr("%s contains '..' traversal: %s", source, path)
	}

	resolved, err := filepath.Abs(path)
	if err != nil {
		return "", fmtErr("%s: resolve path: %w", source, err)
	}

	// Block system directories
	if IsBlockedPath(resolved) {
		return "", fmtErr("%s targets a system directory: %s", source, resolved)
	}

	// Must not be a symlink itself (prevents symlink escalation)
	// Use Lstat to not follow symlinks
	if info, err := os.Lstat(path); err == nil {
		if info.Mode()&os.ModeSymlink != 0 {
			return "", fmtErr("%s is a symlink (not allowed): %s", source, path)
		}
	} else if !os.IsNotExist(err) {
		// Permission denied or other stat error
		return "", fmtErr("%s cannot be accessed: %w", source, err)
	}

	// Must not exceed a reasonable path depth
	parts := strings.Split(resolved, string(filepath.Separator))
	// Filter empty parts (leading /)
	nonEmpty := 0
	for _, p := range parts {
		if p != "" {
			nonEmpty++
		}
	}
	if nonEmpty > maxPathDepth {
		return "", fmtErr("%s path too deep (%d components): %s", source, nonEmpty, resolved)
	}

	return resolved, nil
}

// ValidateWritePath validates that a write target path is safe.
// Same checks as ValidateHomePath but for write targets.
// Does NOT require the path to be within the LLMem home directory —
// users may configure custom output paths.
func ValidateWritePath(path, source string) (string, error) {
	// Check traversal BEFORE resolving
	if strings.Contains(filepath.Clean(path), "..") {
		return "", fmtErr("%s path contains '..' traversal: %s", source, path)
	}

	resolved, err := filepath.Abs(path)
	if err != nil {
		return "", fmtErr("%s: resolve path: %w", source, err)
	}

	// Block system directories
	if IsBlockedPath(resolved) {
		return "", fmtErr("%s path targets a protected directory: %s", source, resolved)
	}

	// Must not be a symlink itself
	if info, err := os.Lstat(path); err == nil {
		if info.Mode()&os.ModeSymlink != 0 {
			return "", fmtErr("%s path is a symlink (not allowed for write targets): %s", source, path)
		}
	} else if !os.IsNotExist(err) {
		return "", fmtErr("%s path cannot be accessed: %w", source, err)
	}

	return resolved, nil
}

// ValidateSessionID validates that a session ID is safe to use in filesystem paths.
// Rejects session IDs that contain path separators or traversal sequences.
func ValidateSessionID(sessionID string) (string, error) {
	if sessionID == "" {
		return "", fmtErr("session_id must not be empty")
	}
	if strings.Contains(sessionID, "/") {
		return "", fmtErr("session_id contains '/' (path traversal risk): %q", sessionID)
	}
	if strings.Contains(sessionID, "\\") {
		return "", fmtErr("session_id contains '\\' (path traversal risk): %q", sessionID)
	}
	if strings.Contains(sessionID, "..") {
		return "", fmtErr("session_id contains '..' (path traversal risk): %q", sessionID)
	}
	return sessionID, nil
}

// MigrateFromLobsterdog migrates data from ~/.lobsterdog/ to ~/.config/llmem/.
// Copies config.yaml, memory.db, dream_diary.md, proposed-changes.md,
// and context/ directory from ~/.lobsterdog/ to ~/.config/llmem/
// only if the source exists and the destination doesn't.
// Never deletes the source directory.
//
// Returns (true, nil) if any migration was performed, (false, nil) if no migration needed.
func MigrateFromLobsterdog() (bool, error) {
	homeDir, err := os.UserHomeDir()
	if err != nil {
		return false, fmtErr("get home directory: %w", err)
	}

	oldHome := filepath.Join(homeDir, ".lobsterdog")
	newHome := filepath.Join(homeDir, ".config", "llmem")

	// Resolve to avoid symlink attacks
	oldHome, err = filepath.Abs(oldHome)
	if err != nil {
		return false, fmtErr("resolve old home: %w", err)
	}
	newHome, err = filepath.Abs(newHome)
	if err != nil {
		return false, fmtErr("resolve new home: %w", err)
	}

	if _, err := os.Stat(oldHome); os.IsNotExist(err) {
		return false, nil
	}
	if _, err := os.Stat(newHome); err == nil {
		return false, nil
	}

	migrated := false
	if err := os.MkdirAll(newHome, 0700); err != nil {
		return false, fmtErr("create new home directory: %w", err)
	}

	filesToCopy := []string{
		"config.yaml",
		"memory.db",
		"dream_diary.md",
		"proposed-changes.md",
	}

	for _, filename := range filesToCopy {
		src := filepath.Join(oldHome, filename)
		dst := filepath.Join(newHome, filename)

		// Skip symlinks — only copy regular files
		if info, err := os.Lstat(src); err == nil {
			if info.Mode()&os.ModeSymlink != 0 {
				slog.Warn("llmem: paths: skipping symlink during migration", "path", src)
				continue
			}
		}

		if _, err := os.Stat(src); err != nil {
			continue // Source doesn't exist
		}
		if _, err := os.Stat(dst); err == nil {
			continue // Destination already exists
		}

		if err := copyFile(src, dst); err != nil {
			slog.Warn("llmem: paths: failed to migrate file", "file", filename, "error", err)
			continue
		}
		slog.Info("llmem: paths: migrated file from ~/.lobsterdog/ to ~/.config/llmem/", "file", filename)
		migrated = true
	}

	// Copy context/ directory — skip symlinks
	srcContext := filepath.Join(oldHome, "context")
	dstContext := filepath.Join(newHome, "context")

	if info, err := os.Lstat(srcContext); err == nil {
		if info.Mode()&os.ModeSymlink != 0 {
			slog.Warn("llmem: paths: skipping symlink directory during migration", "path", srcContext)
		} else if info.IsDir() {
			if _, err := os.Stat(dstContext); os.IsNotExist(err) {
				if err := copyDir(srcContext, dstContext); err != nil {
					slog.Warn("llmem: paths: failed to migrate context/ directory", "error", err)
				} else {
					slog.Info("llmem: paths: migrated context/ directory from ~/.lobsterdog/ to ~/.config/llmem/")
					migrated = true
				}
			}
		}
	}

	return migrated, nil
}

// copyFile copies a single file from src to dst.
func copyFile(src, dst string) error {
	data, err := os.ReadFile(src)
	if err != nil {
		return fmtErr("read source file: %w", err)
	}
	return os.WriteFile(dst, data, 0600)
}

// copyDir copies a directory tree from src to dst.
func copyDir(src, dst string) error {
	return filepath.Walk(src, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}

		// Skip symlinks
		if info.Mode()&os.ModeSymlink != 0 {
			return nil
		}

		relPath, err := filepath.Rel(src, path)
		if err != nil {
			return err
		}
		targetPath := filepath.Join(dst, relPath)

		if info.IsDir() {
			return os.MkdirAll(targetPath, 0700)
		}

		return copyFile(path, targetPath)
	})
}